File: src/android/src/com/che/zap/play/ZwiftPlayDevice.kt
package com.che.zap.play

import com.che.zap.device.ZapConstants.BATTERY_LEVEL_TYPE
import com.che.zap.device.ZapConstants.CONTROLLER_NOTIFICATION_MESSAGE_TYPE
import com.che.zap.device.ZapConstants.EMPTY_MESSAGE_TYPE
import com.che.zap.device.AbstractZapDevice
import com.che.zap.device.ZapConstants.CLICK_TYPE
import com.che.zap.proto.BatteryStatus
import com.che.zap.proto.ControllerNotification
import com.che.zap.utils.Logger
import com.che.zap.utils.toHexString
import timber.log.Timber

class ZwiftPlayDevice : AbstractZapDevice() {

    // you get battery level in a BLE characteristic and via a ZAP message.
    private var batteryLevel = 0

    private var lastButtonState: ControllerNotification? = null

    override fun processEncryptedData(bytes: ByteArray) {
        try {

            if (LOG_RAW) Timber.d("Decrypted: ${bytes.toHexString()}")

            val counter = bytes.copyOfRange(0, Int.SIZE_BYTES)
            val payload = bytes.copyOfRange(Int.SIZE_BYTES, bytes.size)

            val data = zapEncryption.decrypt(counter, payload)
            val type = data[0]
            val message = data.copyOfRange(1, data.size)

            when (type) {
                CONTROLLER_NOTIFICATION_MESSAGE_TYPE -> processButtonNotification(ControllerNotification(message))
                CLICK_TYPE -> processClickButtonNotification(data)
                EMPTY_MESSAGE_TYPE -> if (LOG_RAW) Logger.d("Empty Message") // expected when nothing happening
                BATTERY_LEVEL_TYPE -> {
                    val notification = BatteryStatus(message)
                    if (batteryLevel != notification.level) {
                        batteryLevel = notification.level
                        Logger.d("Battery level update: $batteryLevel")
                    }
                }
                else -> Logger.e("Unprocessed - Type: ${type.toUByte().toHexString()} Data: ${data.toHexString()}")
            }

        } catch (ex: Exception) {
            Logger.e("Decrypt failed: " + ex.message)
        }
    }

    private fun processButtonNotification(notification: ControllerNotification) {
        if (lastButtonState == null)
            Logger.d(notification.toString())
        else {
            val diff = notification.diff(lastButtonState!!)
            if (!diff.isNullOrBlank()) // get repeats of the same state
                Logger.d(diff)
        }
        lastButtonState = notification
    }

    private fun processClickButtonNotification(data: ByteArray) {
        //Logger.d("Click Button Press ${data.toHexString()}")
        if(data.size == 5) {
            if(data[4] == 0.toByte()) {
                Logger.d("Click '-' Button Press")
            } else if(data[2] == 0.toByte()) {
                Logger.d("Click '+' Button Press")
            }
        }
    }
}

File: src/ios/zwift_play/zwiftPlayDevice.swift
import Foundation
import os.log

@available(iOS 14.0, *)
class ZwiftPlayDevice: AbstractZapDevice {
    private var batteryLevel = 0

    override func processEncryptedData(bytes: Data) -> Int {
        do {
            os_log("Decrypted: %{public}@", log: OSLog.default, type: .debug, bytes.toHexString())

            let counter = bytes.prefix(MemoryLayout<Int>.size)
            let payload = bytes.suffix(from: MemoryLayout<Int>.size)

            guard let data = zapEncryption.decrypt(counterArray: counter, payload: payload) else {
                os_log("Decrypt failed", log: OSLog.default, type: .error)
                return 0
            }

            let type = data[0]
            let message = data.suffix(from: 1)

            switch type {
            case ZapConstants.controllerNotificationMessageType:
                os_log("controllerNotificationMessageType")
                //processButtonNotification(notification: ControllerNotification(data: message))
            case ZapConstants.clickType:
                return processClickButtonNotification(data: data)
            case ZapConstants.emptyMessageType:
                os_log("emptyMessageType")
            case ZapConstants.batteryLevelType:
                os_log("batteryLevelType")
            default:
                os_log("Unprocessed - Type: %{public}@ Data: %{public}@", log: OSLog.default, type: .error, String(describing: type), data.toHexString())
            }
        } catch {
            os_log("Decrypt failed: %{public}@", log: OSLog.default, type: .error, error.localizedDescription)
        }
        return 0
    }
/*
    private func processButtonNotification(notification: ControllerNotification) {
        if let lastButtonState = lastButtonState {
            let diff = notification.diff(lastButtonState: lastButtonState)
            if !diff.isEmpty {
                os_log("%{public}@", log: OSLog.default, type: .debug, diff)
            }
        } else {
            os_log("%{public}@", log: OSLog.default, type: .debug, notification.description)
        }
        lastButtonState = notification
    }*/

    private func processClickButtonNotification(data: Data) -> Int {
        if data.count == 5 {
            if data[4] == 0 {
                os_log("Click '-' Button Press", log: OSLog.default, type: .debug)
                return 2
            } else if data[2] == 0 {
                os_log("Click '+' Button Press", log: OSLog.default, type: .debug)
                return 1
            }
        }
        return 0
    }
}
